kola: Replace run-ext-bin with `kola run -E/--exttest`

cgwalters opened the pull request:
As I tried to port the ostree tests, I hit various issues
with the `run-ext-bin` model; for example, `kola run` already
supports `--parallel` and cleanly separates tests into
output directories etc.

This reworks things so that one can do e.g.:

`cosa kola run -- --parallel 4 --output-dir tmp/kola -E ~/src/github/ostreedev/ostree/tests/installed/nondestructive/ 'ext.nondestructive.*'`

to run the ostree test suite.  Or of course, run *both* the
ostree exttests plus builtin tests `basic` or whatever.

And another thing I'd like to do soon is teach cosa how to
automatically add `src/config/tests` so that one can seamlessly
e.g. add a change to the manifest and also add a test for it
as one atomic commit.

There's still a lot more to do here; among other things ostree
cleanly separates destructive/nondestructive tests and having
each test run as a new VM is a hit.  But this is a good start
I think.

cgwalters commented:
I updated https://github.com/ostreedev/ostree/pull/2025 to use this as a demo.

ashcrow commented on mantle/kola/harness.go:
```diff
@@ -436,6 +448,145 @@ func RunUpgradeTests(patterns []string, pltfrm, outputDir string, propagateTestE
+`, kolaExtBinDataEnv, kolaExtBinDataDir, remotepath)
+	runextconfig := ignv3types.Config{
+		Ignition: ignv3types.Ignition{
+			Version: "3.0.0",
```
Will this matter for RHCOS being on spec2x?

cgwalters replied:
Because this just hooks into the main `kola run` harness code, we reuse the [Ignition conversion](https://github.com/coreos/mantle/pull/1186) that already landed.

ashcrow commented on mantle/kola/harness.go:
```diff
@@ -436,6 +448,145 @@ func RunUpgradeTests(patterns []string, pltfrm, outputDir string, propagateTestE
+				return err
+			}
+			subprefix := fmt.Sprintf("%s.%s", testprefix, c.Name())
+			if err := registerTestDir(subdir, subprefix, subchildren); err != nil {
```
recursion FTW :+1:

ashcrow commented on mantle/kola/README-kola-ext.md:
```diff
@@ -10,28 +10,38 @@ repositories, and allow these projects to target Fedora CoreOS
+- `deps` (optional): Directory (or symlink to dir): Will be uploaded and available as `KOLA_EXT_DATA`
+- one or more executables: Each executable is its own test, run independently
+  with the Ignition config and/or dependencies provided.
+
```
I think something referencing the transformation could be helpful. Something like:

```suggestion

*Note*: When adding tests in this way the test prefix comes from the basename of the directory and the test name is the basename of the binary itself. As an example `/path/to/externaltest/tests/mytest` where `-E /path/to/externaltest/` was used would be available in the kola test suite as `externaltest.tests.mytest`.
```

cgwalters replied:
I adapted a variant of this, thanks!

arithx commented on mantle/kola/harness.go:
```diff
@@ -436,6 +448,153 @@ func RunUpgradeTests(patterns []string, pltfrm, outputDir string, propagateTestE
+		}
+	}
+
+	if len(executables) == 1 {
```
Do you think that putting everything in a loop and just special casing the name is more readable, i.e.:

```golang
for _, executable := range executables {
    testname := testprefix
    if len(executables) > 1 {
        testname = fmt.Sprintf("%s.%s", testname, filepath.Base(executable))
    }
    err := registerExternalTest(testname, executable, dependencydir, ignition)
    if err != nil {
        return err
    }
}
```

arithx commented on mantle/kola/register/register.go:
```diff
@@ -52,6 +52,7 @@ type Test struct {
 	Name                 string // should be unique
 	Run                  func(cluster.TestCluster)
 	NativeFuncs          map[string]NativeFuncWrap
+	ExternalTest         string
```
Could use a comment explaining this is the path to the binary.

cgwalters commented:
New demonstration: https://github.com/coreos/fedora-coreos-config/pull/305

cgwalters commented:
I discovered the rebooting support was even more aspirational rather than functional.

arithx commented:
> I discovered the rebooting support was even more aspirational rather than functional.

What specifically are you seeing? I haven't tried it with `NativeFuncs` but it works fine for normal tests.

cgwalters commented:
> What specifically are you seeing? I haven't tried it with NativeFuncs but it works fine for normal tests.

This code is trying to support having the test code run `/usr/sbin/reboot`.  Remember a *huge* difference between these external tests and most of the rest of kola is the idea that the test is a binary running on the target, we just poll its status via ssh.

This avoids a whole lot synchronous round trips with `ssh` and allows writing e.g. a test as a 3 line shell script instead of needing to write Go.  (And yes, I do think shell scripts under 10 lines that don't involve functions can be more expressive and clear than Go code invoking SSH)

But the current kola code around reboots is much more oriented around having kola drive/monitor the reboot status so I'm reworking things to have the test code request a reboot rather than invoke it directly.

cgwalters commented:
OK cool, I fixed up the rebooting support (not in the most elegant way, but eh it works) and also rewrote the `rhcos.luks.tpm` test using this: https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/871

cgwalters commented:
Oh and this now supports a `kola.json` that can be in the test directory, e.g. for that one:

```
{
    "architectures": "!s390x ppc64le",
    "platforms": "qemu-unpriv"
}
```

cgwalters commented:
OK, updated the docs and lifting WIP on this!

cgwalters commented on mantle/cmd/kolet/kolet.go:
```diff
@@ -138,9 +139,16 @@ func dispatchRunExtUnit(unitname string, sdconn *systemddbus.Conn) (bool, error)
 				case CLD_KILLED:
 					// SIGTERM; we explicitly allow that, expecting we're rebooting.
-					if mainstatus == 15 {
+					if mainstatus == int32(15) {
```
Oh yeah this...SO EVIL.

ashcrow commented on mantle/cmd/kolet/kolet.go:
```diff
@@ -138,9 +139,16 @@ func dispatchRunExtUnit(unitname string, sdconn *systemddbus.Conn) (bool, error)
+							Pid: os.Getpid(),
+						}
+						selfproc.Signal(syscall.SIGTERM)
+						time.Sleep(time.Hour)
```
It feels odd to wait wait up to an hour, but then again it won't hurt to wait that long since it should be interrupted.

cgwalters commented:
Another big topic here is that I think we need to standardize a bit more of the flow for "build project X, overlay it on fcos, run *just the tests for that project* (plus possibly kola basic or others) as external tests".   A bit more discussion of related topics in https://github.com/coreos/rpm-ostree/pull/1958

cgwalters commented:
And a new major update to https://github.com/ostreedev/ostree/pull/2025 !

jlebon commented on mantle/kola/README-kola-ext.md:
```diff
@@ -10,31 +13,83 @@ repositories, and allow these projects to target Fedora CoreOS
-Currently this systemd unit runs with full privileges - tests
+- `config.ign` (optional): Ignition config provided
+- `kola.json` (optional): JSON file described below
+- `deps` (optional): Directory (or symlink to dir): Will be uploaded and available as `${KOLA_EXT_DATA}`
```
This is bikeshed, though WDYT about `data/` (or maybe even `extdata/`) for this? Goes with `KOLA_EXT_DATA` and other places in the code.

jlebon commented on mantle/kola/README-kola-ext.md:
```diff
@@ -10,31 +13,83 @@ repositories, and allow these projects to target Fedora CoreOS
+
+An important feature of exttests is support for rebooting the host system.
+This allows one to easily test OS updates for example.  To do this, your
+test process should send `SIGTERM` to itself.  For example, in bash use:
```
I'm a little unsure about the `SIGTERM` convention here. It feels a little generic/implicit. Can we have something more explicit for requesting a reboot?

Hmm, how about a designated exit code? E.g. service exits 77, kolet itself exits 77, and kola reboots.

jlebon replied:
(BTW, I really like that we're tackling the reboot problem here; this is something that the rpm-ostree testsuite will benefit from too!)

cgwalters replied:
I went with an exit code at first but, if we want to support having the test just invoke `reboot`, it's going to get `SIGTERM`, so I think this approach is better.  It feels pretty explicit to me.

jlebon replied:
Right, I get that that's what happens anyway during a reboot. It just feels odd to go from there and flip this around to assume that if that's what happened, then it must mean we're rebooting. Anyway, I'm cool rolling with this to start with.

jlebon commented on mantle/kola/harness.go:
```diff
@@ -152,6 +160,12 @@ var (
 )
 
+// kolaExtBinDataDir is where data will be stored
+const kolaExtBinDataDir = "/var/opt/kola/extdata"
```
Should this be suffixed by the test name? Would make it easier to run multiple tests in parallel in the future.

cgwalters replied:
Running tests in parallel on the same machine conflicts with the idea of destructive tests (which, as far as kola knows all tests are today).

One can already today drop multiple executables in the same dir, have them share the same data, and do "internal parallelization" at least.

arithx replied:
Technically the `NativeFuncs` are considered non-destructive and run on the same machine but they get reported as sub-tests. I had a PR forever ago that never ended up landing which worked at adding non-destructive tests (https://github.com/coreos/mantle/pull/697) if someone ever wants to pick that up and re-work it to fit modern mantle.

jlebon commented:
We already ahve two :heavy_check_mark: reviews on this, so
/lgtm
